File: api/services/rabbitmq/publisher.go
package services

import (
	"context"
	"log"

	amqp "github.com/rabbitmq/amqp091-go"
)

type RabbitMqPublisher struct{}

func (pub *RabbitMqPublisher) Publish(ctx context.Context, queue string, message string) {
	conn, err := amqp.Dial("amqp://guest:guest@rabbitmq:5672/")
	if err != nil {
		log.Fatal(err)
	}

	ch, err := conn.Channel()
	if err != nil {
		log.Fatal(err)
	}

	q, err := ch.QueueDeclare(queue, false, false, false, false, nil)
	if err != nil {
		log.Fatal(err)
	}

	err = ch.PublishWithContext(ctx, "", q.Name, false, false, amqp.Publishing{
		ContentType: "text/plain",
		Body:        []byte(message),
	})
	if err != nil {
		log.Fatal(err)
	}
}

File: parser/src/services/rabbitmq/publisher.py
import pika
import pika.adapters.blocking_connection
import pika.channel


class RabbitMqPublisher:
    _self = None
    _connection: pika.BlockingConnection | None = None
    _channel: pika.adapters.blocking_connection.BlockingChannel | None = None

    def __new__(cls):
        if not cls._self:
            cls._self = super().__new__(cls)
            cls._connection = pika.BlockingConnection(
                pika.ConnectionParameters("rabbitmq"),
            )
            cls._channel = cls._connection.channel()
        return cls._self

    def publish(self, queue: str, message: any) -> None:
        self._channel.queue_declare(queue=queue)
        self._channel.basic_publish(
            exchange="",
            routing_key=queue,
            body=message,
        )
